# alphafold-viral-protein-analysis
This Python-based pipeline automates the retrieval and codon-aware translation of viral FASTA sequences from accession IDs, producing structured JSON outputs, AlphaFold-compatible protein inputs, and 3D structural models (PDB) for downstream analysis in AlphaFold and ChimeraX. The pipeline also generates a residue-level biological report—including hydrophobicity, polarity, folding propensity, and conserved motifs—enabling rapid exploratory analysis of viral genome segments and prediction confidence metrics (pLDDT).

## Biological Motivation
Many virus structures encode multifunctional proteins that are difficult to compare and analyze due to time constraints and experimental challenges. Computational structure analysis provides both baseline and hypothesized structural analysis for functional domains, disorder regions, and potential interfaces prior to wet-lab validation. 

## Methods and Pipeline
1. Viral nucleotide FASTA sequence is retrieved via accession ID input
2. Sequences are read and translated by codon-aware reading frames
3. Protein FASTA, JSON, PDB, and A3M files are generated into correlating subfolders
4. Structural predictions are run and generated by ColabFold
5. Output confidence metrics (pLDDT) are summarized and reported
6. A structural analysis recognizing amino acid interactions (ie. hydrophobicity, polarity, conserved motifs, and folding propensity) is generated

PIPELINE: Prompts accession ID --> Search library for FASTA nucleotide sequence --> Codon-aware translation --> Baseline + Hypothesized JSON and FASTA --> ColabFold --> pLDDT + PDB --> Manual structural analysis 

### Example Output
KU343156 Farallon virus isolate CalAr846 cRNA segment S
python combined_tools.py --accession KU343156
###Predicted Protein Structure 
<img width="554" height="899" alt="376D848E94F34279" src="https://github.com/user-attachments/assets/fb22d216-1f37-4661-bd4e-e84b212b872f" />

<img width="1347" height="939" alt="KU343156_virus__Farallon_virus_isolate_CalAr846_cRNA_segment___S_full_length___494_coverage" src="https://github.com/user-attachments/assets/e81bb918-9ea9-4e47-92e4-df3f83b43d01" />
High pLDDT regions (>90) are considered structurally reliable, while low-confidence regions may indicate intrinsic disorder or flexible linkers.

<img width="2435" height="438" alt="KU343156_virus__Farallon_virus_isolate_CalAr846_cRNA_segment___S_full_length___494_pae" src="https://github.com/user-attachments/assets/3d237402-b94f-45e6-b519-45370f0875ab" />

<img width="1391" height="939" alt="KU343156_virus__Farallon_virus_isolate_CalAr846_cRNA_segment___S_full_length___494_plddt" src="https://github.com/user-attachments/assets/67e69420-c390-4bcf-898b-6a22e7c4a11b" />

<img width="1102" height="810" alt="Screenshot 2026-01-19 105619" src="https://github.com/user-attachments/assets/e68939e8-16f1-417c-bee0-7f5cd75680e1" />


### Output Visualization  

Accession: KU343156
Virus: Farallon virus isolate CalAr846 cRNA
Segment: S
------------------------------------------------------------------------------------------------------------------------
[ORF] accession = KU343156 start_nt=87 frame = 2 nt_len=1482 aa_len=494
[AlphaFoldManager] Running: /home/katie/localcolabfold/localcolabfold/colabfold-conda/bin/colabfold_batch colabfold_results/KU343156/KU343156.fasta colabfold_results/KU343156
2026-01-18 20:25:08,390 Running colabfold 1.5.5 (83ee93d262a99ad62d6f0897c5ddd37eb918d385)

WARNING: You are welcome to use the default MSA server, however keep in mind that it's a
limited shared resource only capable of processing a few thousand MSAs per day. Please
submit jobs only from a single IP address. We reserve the right to limit access to the
server case-by-case when usage exceeds fair use. If you require more MSAs: You can
precompute all MSAs with `colabfold_search` or host your own API and pass it to `--host-url`

2026-01-18 20:25:08,670 WARNING: no GPU detected, will be using CPU
2026-01-18 20:25:09,051 Found 5 citations for tools or databases
2026-01-18 20:25:09,058 Query 1/1: KU343156_virus__Farallon_virus_isolate_CalAr846_cRNA_segment___S_full_length___494 (length 494)
2026-01-18 20:25:09,991 Setting max_seq=148, max_extra_seq=1
WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
I0000 00:00:1768796738.206317    7511 mlir_graph_optimization_pass.cc:437] MLIR V1 optimization pass is not enabled
2026-01-18 20:49:05,675 alphafold2_ptm_model_1_seed_000 recycle=0 pLDDT=82.9 pTM=0.742
2026-01-18 21:12:12,444 alphafold2_ptm_model_1_seed_000 recycle=1 pLDDT=87.5 pTM=0.782 tol=4.75
2026-01-18 21:45:55,748 alphafold2_ptm_model_1_seed_000 recycle=2 pLDDT=87.8 pTM=0.777 tol=0.473
2026-01-18 22:10:56,352 alphafold2_ptm_model_1_seed_000 recycle=3 pLDDT=87.9 pTM=0.775 tol=0.228
2026-01-18 22:10:56,402 alphafold2_ptm_model_1_seed_000 took 6316.2s (3 recycles)
2026-01-18 22:28:24,324 alphafold2_ptm_model_2_seed_000 recycle=0 pLDDT=83.9 pTM=0.75
2026-01-18 22:45:33,668 alphafold2_ptm_model_2_seed_000 recycle=1 pLDDT=87.1 pTM=0.767 tol=1.88
2026-01-18 23:02:43,967 alphafold2_ptm_model_2_seed_000 recycle=2 pLDDT=86.9 pTM=0.762 tol=0.508
2026-01-18 23:19:54,404 alphafold2_ptm_model_2_seed_000 recycle=3 pLDDT=87.1 pTM=0.765 tol=1.2
2026-01-18 23:19:54,433 alphafold2_ptm_model_2_seed_000 took 4137.4s (3 recycles)
2026-01-18 23:37:01,452 alphafold2_ptm_model_3_seed_000 recycle=0 pLDDT=86.5 pTM=0.771
2026-01-18 23:54:12,326 alphafold2_ptm_model_3_seed_000 recycle=1 pLDDT=88.6 pTM=0.785 tol=1.03
2026-01-19 00:11:28,893 alphafold2_ptm_model_3_seed_000 recycle=2 pLDDT=89.2 pTM=0.792 tol=0.449
2026-01-19 00:28:42,845 alphafold2_ptm_model_3_seed_000 recycle=3 pLDDT=89.5 pTM=0.802 tol=0.264
2026-01-19 00:28:42,904 alphafold2_ptm_model_3_seed_000 took 4127.8s (3 recycles)
2026-01-19 00:46:00,942 alphafold2_ptm_model_4_seed_000 recycle=0 pLDDT=86.8 pTM=0.772
2026-01-19 01:03:16,402 alphafold2_ptm_model_4_seed_000 recycle=1 pLDDT=88.6 pTM=0.771 tol=1.42
2026-01-19 01:20:28,734 alphafold2_ptm_model_4_seed_000 recycle=2 pLDDT=88.5 pTM=0.767 tol=0.676
2026-01-19 01:37:43,206 alphafold2_ptm_model_4_seed_000 recycle=3 pLDDT=88.7 pTM=0.765 tol=0.345
2026-01-19 01:37:43,238 alphafold2_ptm_model_4_seed_000 took 4139.7s (3 recycles)
2026-01-19 01:55:00,564 alphafold2_ptm_model_5_seed_000 recycle=0 pLDDT=85.8 pTM=0.771
2026-01-19 02:12:26,218 alphafold2_ptm_model_5_seed_000 recycle=1 pLDDT=87.8 pTM=0.794 tol=1.33
2026-01-19 02:29:49,468 alphafold2_ptm_model_5_seed_000 recycle=2 pLDDT=87.8 pTM=0.786 tol=0.43
2026-01-19 02:47:07,289 alphafold2_ptm_model_5_seed_000 recycle=3 pLDDT=87.9 pTM=0.792 tol=0.775
2026-01-19 02:47:07,335 alphafold2_ptm_model_5_seed_000 took 4163.6s (3 recycles)
2026-01-19 02:47:07,862 reranking models by 'plddt' metric
2026-01-19 02:47:07,874 rank_001_alphafold2_ptm_model_3_seed_000 pLDDT=89.5 pTM=0.802
2026-01-19 02:47:07,906 rank_002_alphafold2_ptm_model_4_seed_000 pLDDT=88.7 pTM=0.765
2026-01-19 02:47:07,929 rank_003_alphafold2_ptm_model_5_seed_000 pLDDT=87.9 pTM=0.792
2026-01-19 02:47:07,954 rank_004_alphafold2_ptm_model_1_seed_000 pLDDT=87.9 pTM=0.775
2026-01-19 02:47:07,970 rank_005_alphafold2_ptm_model_2_seed_000 pLDDT=87.1 pTM=0.765
2026-01-19 02:47:13,771 Done
[AlphaFoldManager] ColabFold finished with return code: 0
[AlphaFoldTools]
Mode: ColabFold
Ran ColabFold on KU343156 (494 aa).
FASTA file: colabfold_results/KU343156/KU343156.fasta
Results directory: colabfold_results
------------------------------------------------------------------------------------------------------------------------
  High pLDDT: 624
  Low pLDDT:  8
------------------------------------------------------------------------------------------------------------------------
[Sequence / Structure Analysis Report]
--- Structural Summary ---
File: KU343156
Total Residues: 632
Peptide Preview: XXQRKTXQTXPXHXNXXXXXRRXGKHXXRXQXAXXXRXXXXXXKXXAXGXXXXXXXRPNX...

Motifs:
Basic cluster at position 87
RNA-binding at position 207

Secondary Structure Summary:
-- Alpha Helices: 115 residues
-- Beta-Sheets: 46 residues
-- Coils/Loops: 471 residues

Disordered Regions:
-- Disordered region 1: 259-266 (Score: 37.50)

Overall plDDT Confidence
-- High Confidence (>90) residues: 624
-- Low Confidence (<50) residues: 8

Hydrophobic Core Residues (indices):
10, 23, 32, 46, 48, 57, 64, 73, 93, 114, 116, 122, 148, 150, 188, 198, 201, 216, 218, 225, 232, 234, 255, 258, 272, 274, 328, 331, 353, 355, 375, 382, 395, 396, 404, 411, 422, 423, 433, 454, 469, 470, 489, 504, 509, 564, 585, 624, 625, 628

--- Domain Analysis ---
Domains:
Domain 1 (Residues 1-259)
Domain 2 (Residues 267-632)
========================================================================================================================

## Features
- Prompts the input of accession ID or "QUIT" initiating a loop
- Retrieves nucleotide sequence from library outputting virus name and segment
- Translates nucleotide sequence into correct amino acid sequence and count
- Runs manual ColabFold or AlphaFold3
- Generates JSON and PDB file for 3D imaging
- Computes and summarizes pLDDT confidence charts for the predicted ranks
- Constructs subfolder of given PNGs, JSONs, PDBs, and FASTAs titled with accession ID
- Uses biological principles to output structure analysis based on amino acid composition

## Tech Stack 
- Programming: Python 
- AlphaFold, ColabFold
- Jupyter, Jupyter Notebook, Linux(Ubuntu)
- FASTA, PDB, JSON, PNG

## Validation and Limitatons
## How to Run
1. Copy and download pipeline
2. Pip install ColabFold and requirements.txt
3. Run pipeline
   - cd ~/localcolabfold
     source localcolabfold/conda/etc/profile.d/conda.sh
     conda activate ~/localcolabfold/localcolabfold.colabfold-conda
     export JAX_PLATFORMS=cpu
     cd/mnt/c/Users/"user"/alphafold_project

   - cd/mnt/c/Users/"user"/alphafold_project
     rm -rf __pycache__

   - python combined_tools.py \
      --fasta data/sequences.fasta \
      --colabfold-exe/home/"user"/localcolabfold/localcolabfold/colabfold-conda/bin/colabfold_batch
## File outputs
- PDB files for predicted protein structures
- Manual Baseline JSONs for AlphaFold
- JSON files with confidence metrics
- pLDDT plots
- Text-based biophysical summaries

## Skills Demonstrated
- Viral genomics
- Protein structure prediction
- Structural confidence interpretation
- Computational biology pipeline design

  
## Next Steps
- Continue to expand the reference library to make it more widely functional
- Extend framework to enzymes, structural proteins, and other classes by additionally expanding the library and creating more if/elif/else attributes for specific analysis
- Integrating additional biotech tools to improve visual analysis and analytical depth
- Additional dynamics and domain comparisons
- Creating additional hypothetical JSON files for postranslational modifications 
