# alphafold-viral-protein-analysis
This Python-based pipeline automates the retrieval and codon-aware translation of viral FASTA sequences from accession IDs, producing structured JSON outputs, AlphaFold-compatible protein inputs, and 3D structural models (PDB) for downstream analysis in AlphaFold and ChimeraX. The pipeline also generates a residue-level biological report—including hydrophobicity, polarity, folding propensity, and conserved motifs—enabling rapid exploratory analysis of viral genome segments and prediction confidence metrics (pLDDT).

## Biological Motivation
Many virus structures encode multifunctional proteins that are difficult to compare and analyze due to time constraints and experimental challenges. Computational structure analysis provides both baseline and hypothesized structural analysis for functional domains, disorder regions, and potential interfaces prior to wet-lab validation. 

## Methods and Pipeline
1. Viral nucleotide FASTA sequence is retrieved via accession ID input
2. Sequences are read and translated by codon-aware reading frames
3. Protein FASTA, JSON, PDB, and A3M files are generated into correlating subfolders
4. Structural predictions are run and generated by ColabFold
5. Output confidence metrics (pLDDT) are summarized and reported
6. A structural analysis recognizing amino acid interactions (ie. hydrophobicity, polarity, conserved motifs, and folding propensity) is generated

PIPELINE: Prompts accession ID --> Search library for FASTA nucleotide sequence --> Codon-aware translation --> Baseline + Hypothesized JSON and FASTA --> ColabFold --> pLDDT + PDB --> Manual structural analysis 

### Example Output
KU343156 Farallon virus isolate CalAr846 cRNA segment S
python combined_tools.py --accession KU343156
### Predicted Protein Structure 
<img width="554" height="899" alt="376D848E94F34279" src="https://github.com/user-attachments/assets/fb22d216-1f37-4661-bd4e-e84b212b872f" />

<img width="1347" height="939" alt="KU343156_virus__Farallon_virus_isolate_CalAr846_cRNA_segment___S_full_length___494_coverage" src="https://github.com/user-attachments/assets/e81bb918-9ea9-4e47-92e4-df3f83b43d01" />
High pLDDT regions (>90) are considered structurally reliable, while low-confidence regions may indicate intrinsic disorder or flexible linkers.

<img width="2435" height="438" alt="KU343156_virus__Farallon_virus_isolate_CalAr846_cRNA_segment___S_full_length___494_pae" src="https://github.com/user-attachments/assets/3d237402-b94f-45e6-b519-45370f0875ab" />

<img width="1391" height="939" alt="KU343156_virus__Farallon_virus_isolate_CalAr846_cRNA_segment___S_full_length___494_plddt" src="https://github.com/user-attachments/assets/67e69420-c390-4bcf-898b-6a22e7c4a11b" />

<img width="1102" height="810" alt="Screenshot 2026-01-19 105619" src="https://github.com/user-attachments/assets/e68939e8-16f1-417c-bee0-7f5cd75680e1" />


## Output Visualization  

  Accession: KU343156
  Virus: Farallon virus isolate CalAr846 cRNA
  Segment: S
------------------------------------------------------------------------------------------------------------------------
[ORF] accession = KU343156 start_nt=87 frame = 2 nt_len=1482 aa_len=494
[AlphaFoldManager] Running: /home/katie/localcolabfold/localcolabfold/colabfold-conda/bin/colabfold_batch colabfold_results/KU343156/KU343156.fasta colabfold_results/KU343156
2026-01-18 20:25:08,390 Running colabfold 1.5.5 (83ee93d262a99ad62d6f0897c5ddd37eb918d385)

****[AlphaFoldManager] ColabFold finished with return code: 0
[AlphaFoldTools]
Mode: ColabFold
Ran ColabFold on KU343156 (494 aa).
FASTA file: colabfold_results/KU343156/KU343156.fasta
Results directory: colabfold_results****
------------------------------------------------------------------------------------------------------------------------
  High pLDDT: 624
  Low pLDDT:  8
------------------------------------------------------------------------------------------------------------------------
[Sequence / Structure Analysis Report]
--- Structural Summary ---
File: KU343156
Total Residues: 632
Peptide Preview: XXQRKTXQTXPXHXNXXXXXRRXGKHXXRXQXAXXXRXXXXXXKXXAXGXXXXXXXRPNX...

Motifs:
Basic cluster at position 87
RNA-binding at position 207

Secondary Structure Summary:
-- Alpha Helices: 115 residues
-- Beta-Sheets: 46 residues
-- Coils/Loops: 471 residues

Disordered Regions:
-- Disordered region 1: 259-266 (Score: 37.50)

Overall plDDT Confidence
-- High Confidence (>90) residues: 624
-- Low Confidence (<50) residues: 8

Hydrophobic Core Residues (indices):
10, 23, 32, 46, 48, 57, 64, 73, 93, 114, 116, 122, 148, 150, 188, 198, 201, 216, 218, 225, 232, 234, 255, 258, 272, 274, 328, 331, 353, 355, 375, 382, 395, 396, 404, 411, 422, 423, 433, 454, 469, 470, 489, 504, 509, 564, 585, 624, 625, 628

--- Domain Analysis ---
Domains:
Domain 1 (Residues 1-259)
Domain 2 (Residues 267-632)

# Cases too large for ColabFold
A primary limitation of this project arises from ColabFold’s sequence token limitations. For inputs that exceed this threshold, the pipeline transitions to a manual AlphaFold execution, producing baseline JSON outputs and enabling continued structural and functional analysis without reliance on ColabFold.

## Example Manual AlphaFold Output 
Accession: MT153554
Virus: Red goblin roach virus 1 isolate OKIAV321 cRNA
Segment: L
------------------------------------------------------------------------------------------------------------------------
[ORF] accession = MT153554 start_nt=62 frame = 1 nt_len=12045 aa_len=4015
[AlphaFoldTools]
Mode: manual_af3
Sequence MT153554 translated to protein is 4015 aa,
which exceeds the automated limit of 1400 aa.

Copy and Paste the FASTA below into AlphaFold3 web:

>MT153554_1_1400 virus: Red_goblin_roach_virus_1_isolate_OKIAV321_cRNA segment = L residues=1-1400 AlphaFold3_manual_input_protein
MEMEVTGTGTEPNSLGSNNDPGESSVEKNVPAVGFIRDIFIFHMDFFLNLSTNVIRLGVETDMQADEIKI
KGRRLIEELRKSNRPVMYRLIDSQMISTIYVLEGILNHKNWMSMIDNYHSQTLKLVGITTYILGSILLAA
RPEDEYIVDMMVVQVGRPELLSVFNKILTDWKAFTKRLKEMIDILYGDRHRKIIGIIQACPKTMLDVTSV
MEIRKVHSVDKWETIMLETAGAYTVNQATNNEKVNNLLKVVKEVKNIFYGPEYSSPNREVQFESRRQLSK
VIENYHNSQQVSKSLREFGIVSSANTISSLVTEVIRALDGGDNSREDSLSLGSLKSELMQMIDSSKQWPP
HVMFRILQRDLEPLFEVISPSYKQLLDDSISRSIMSASYPEAWNNAVKIKGLLYEICFAETFGLPYAPEV
SKLTLNQIVHILYPNAATTFLRRSAEKHMIRDYKPDFFVPYDPIRLEPRRPRFTRPNLTKSAERFYKTGQ
TEPVGPEEDENPVIKWLKNTVDINTKLARLLREVTTARERKYKFNRGGVEDKILSDQEVFSKSGLVIIEV
GYQTSSESKTIIDSRKWKAAANILSSCGIDCTILTLSESTQEMKVDSWLRKDLSTLVSKAVGSLFSKLIA
TMPPSLSNILVGDISSQQLRSLSSEFRPKTCPVSKGDLIDCHVQNKAKLFIRPNGTHLPLHFQRKFVKSL
ITGSLIKESDAHRVKEHLSRNATSIIESVDQSNLCLTKVDDCSPYLFYYSYMKEELESVLERTNVCHELK
KLASECNTACKQLINCLKKEDLSDIDASKFKLRTSLSNFIIRKGHHRSIVLHNCGVSEADRCTIPAIFRL
LHLCGTKKIQVRYSKDDRYVDEVTKSSIDSVSSLTLPGRNAKQKETKGIIRSILETISGLSQDGFLKTES
GLLLSTHKDILLKSNCITGVETGVEKVQVEEEYKIREESTSSQAAPRERKATTVKKQILEKLDSVLKRRD
CLTEEAVESLSRMREDIRSTRVTESDKGHQLFKQLRVSSYIHGQALLDSMHLLLFDSVHTDPLKIFFKGK
EEKAKQMKLKEERESDYRLVSDNMLSDHLSLCESLTEILENETEIEIIPEVVLKQVRESIQPTTPAGVAW
ESAQKALDIIRHTRLGRALRLYSYICQTFLIAVSDMSFGGIKVLKVMNSSTNILIQVGKSKATNCPCILM
DEGMNSITPRFYLNRSVATLGQSLMACVFTNFVQLMQTSSCFDLLSMPLSDILVKVAEHLPEAVTRIGKF
ISDSIRQGYNESTAESLYGSGTMPWELPKPEGQIDRQSQLIACHLACLCFVIVPGVIRNSRKDNKVLQMV
RHPYMLSLSDFGFPLGVGAKLSSPIRNTCTLVLSRFLAQVCCINLEFAKPKLEEWTKSNWRPTSNTVSIS

(Metadata + FASTA also saved to colabfold_results/MT153554/MT153554_af3_manual.json)

(Baseline AF-Server JSON saved to colabfold_results/MT153554/MT153554_baseline.afserver.json)
------------------------------------------------------------------------------------------------------------------------
  High pLDDT: 4183
  Low pLDDT:  0
------------------------------------------------------------------------------------------------------------------------
[Sequence / Structure Analysis Report]
--- Structural Summary ---
File: MT153554
Total Residues: 4183
Peptide Preview: XPRATKXXPXXRXXXRXXXQXGXGXXRHRDRTXXXGXXXRXRXXXXXKEXXXXRXXQGXX...

Motifs:
N-glycosylation at position 1133
Phosphorylation at position 731
Phosphorylation at position 974
Phosphorylation at position 1351
Phosphorylation at position 1584
Basic cluster at position 1528
Basic cluster at position 3884
RNA-binding at position 159
RNA-binding at position 247
RNA-binding at position 557
RNA-binding at position 844
RNA-binding at position 904
RNA-binding at position 1080
RNA-binding at position 1106
RNA-binding at position 1513
RNA-binding at position 1863
RNA-binding at position 1935
RNA-binding at position 2048
RNA-binding at position 2699
RNA-binding at position 2898
RNA-binding at position 2928
RNA-binding at position 2932
RNA-binding at position 2955
RNA-binding at position 3044
RNA-binding at position 3048
RNA-binding at position 3093
RNA-binding at position 3558
RNA-binding at position 3674
RNA-binding at position 3703
RNA-binding at position 3993
RNA-binding at position 4017
RNA-binding at position 4021
RNA-binding at position 4080
RNA-binding at position 4095
Protease cleavage sight at position 56
Protease cleavage sight at position 287
Protease cleavage sight at position 419
Protease cleavage sight at position 1026
Protease cleavage sight at position 1537
Protease cleavage sight at position 2029
Protease cleavage sight at position 2343
Protease cleavage sight at position 2473

Secondary Structure Summary:
-- Alpha Helices: 758 residues
-- Beta-Sheets: 121 residues
-- Coils/Loops: 3304 residues

Disordered Regions:
None

Overall plDDT Confidence
-- High Confidence (>90) residues: 4183
-- Low Confidence (<50) residues: 0

Hydrophobic Core Residues (indices):
1, 3, 8, 21, 23, 35, 57, 62, 64, 79, 83, 97, 108, 113, 138, 158, 199, 209, 217, 218, 220, 231, 242, 257, 288, 290, 308, 332, 338, 341, 345, 346, 351, 359, 368, 369, 377, 379, 389, 393, 396, 409, 420, 434, 467, 479, 486, 487, 509, 512, 517, 521, 542, 566, 567, 598, 625, 652, 661, 662, 667, 681, 685, 691, 693, 696, 732, 736, 776, 780, 792, 810, 841, 877, 880, 910, 914, 921, 967, 973, 975, 980, 999, 1004, 1027, 1033, 1044, 1050, 1061, 1068, 1071, 1075, 1097, 1124, 1138, 1140, 1157, 1195, 1197, 1201, 1205, 1210, 1235, 1246, 1257, 1268, 1286, 1287, 1303, 1305, 1307, 1311, 1312, 1315, 1333, 1342, 1352, 1373, 1387, 1399, 1422, 1444, 1446, 1450, 1451, 1454, 1465, 1467, 1469, 1494, 1496, 1510, 1520, 1533, 1536, 1538, 1544, 1551, 1559, 1564, 1579, 1585, 1586, 1589, 1606, 1644, 1653, 1693, 1696, 1742, 1755, 1764, 1790, 1794, 1806, 1813, 1830, 1855, 1864, 1881, 1886, 1925, 1932, 1949, 1951, 1955, 1972, 1989, 2020, 2026, 2027, 2030, 2049, 2071, 2081, 2083, 2084, 2091, 2095, 2102, 2112, 2120, 2141, 2154, 2179, 2239, 2246, 2256, 2273, 2277, 2279, 2289, 2334, 2338, 2339, 2344, 2374, 2380, 2393, 2394, 2434, 2441, 2444, 2474, 2479, 2486, 2492, 2494, 2495, 2498, 2499, 2501, 2523, 2534, 2560, 2576, 2582, 2606, 2611, 2623, 2632, 2633, 2634, 2641, 2646, 2660, 2675, 2686, 2707, 2717, 2729, 2738, 2741, 2746, 2750, 2756, 2767, 2768, 2808, 2810, 2831, 2839, 2843, 2854, 2857, 2868, 2872, 2896, 2902, 2918, 2921, 2939, 2985, 3009, 3036, 3043, 3063, 3065, 3144, 3162, 3178, 3183, 3186, 3192, 3193, 3233, 3238, 3248, 3277, 3286, 3291, 3307, 3316, 3317, 3328, 3333, 3355, 3356, 3357, 3369, 3374, 3375, 3383, 3392, 3394, 3395, 3401, 3435, 3446, 3452, 3458, 3489, 3508, 3509, 3510, 3511, 3573, 3599, 3604, 3610, 3617, 3625, 3628, 3645, 3648, 3675, 3716, 3752, 3774, 3806, 3828, 3836, 3852, 3870, 3888, 3917, 3934, 3952, 3955, 3970, 3974, 3980, 3989, 4004, 4005, 4006, 4023, 4037, 4038, 4040, 4046, 4056, 4069, 4074, 4091, 4123, 4181, 4182

--- Domain Analysis ---
Domains:
Domain 1 (Residues 1-4183)
========================================================================================================================

## Features
- Prompts the input of accession ID or "QUIT" initiating a loop
- Retrieves nucleotide sequence from library outputting virus name and segment
- Translates nucleotide sequence into correct amino acid sequence and count
- Runs manual ColabFold or AlphaFold3
- Generates JSON and PDB file for 3D imaging
- Computes and summarizes pLDDT confidence charts for the predicted ranks
- Constructs subfolder of given PNGs, JSONs, PDBs, and FASTAs titled with accession ID
- Uses biological principles to output structure analysis based on amino acid composition

## Tech Stack 
- Programming: Python 
- AlphaFold, ColabFold
- Jupyter, Jupyter Notebook, Linux(Ubuntu)
- FASTA, PDB, JSON, PNG

## Validation and Limitatons
## How to Run
1. Copy and download pipeline
2. Pip install ColabFold and requirements.txt
3. Run pipeline
   - cd ~/localcolabfold
     source localcolabfold/conda/etc/profile.d/conda.sh
     conda activate ~/localcolabfold/localcolabfold.colabfold-conda
     export JAX_PLATFORMS=cpu
     cd/mnt/c/Users/"user"/alphafold_project

   - cd/mnt/c/Users/"user"/alphafold_project
     rm -rf __pycache__

   - python combined_tools.py \
      --fasta data/sequences.fasta \
      --colabfold-exe/home/"user"/localcolabfold/localcolabfold/colabfold-conda/bin/colabfold_batch
## File outputs
- PDB files for predicted protein structures
- Manual Baseline JSONs for AlphaFold
- JSON files with confidence metrics
- pLDDT plots
- Text-based biophysical summaries

## Skills Demonstrated
- Viral genomics
- Protein structure prediction
- Structural confidence interpretation
- Computational biology pipeline design

  
## Next Steps
- Continue to expand the reference library to make it more widely functional
- Extend framework to enzymes, structural proteins, and other classes by additionally expanding the library and creating more if/elif/else attributes for specific analysis
- Integrating additional biotech tools to improve visual analysis and analytical depth
- Additional dynamics and domain comparisons
- Creating additional hypothetical JSON files for postranslational modifications 
